topic/types.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topic
+
+import (
+	"github.com/m3db/m3cluster/client"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+)
+
+// Topic defines the topic of messages.
+type Topic interface {
+	// Name returns the name of the topic.
+	Name() string
+
+	// SetName sets the name of the topic.
+	SetName(value string) Topic
+
+	// NumberOfShards returns the total number of shards of the topic.
+	NumberOfShards() uint32
+
+	// SetNumberOfShards sets the total number of shards of the topic.
+	SetNumberOfShards(value uint32) Topic
+
+	// ConsumerServices returns the consumers of the topic.
+	ConsumerServices() []ConsumerService
+
+	// SetConsumerServices sets the consumers of the topic.
+	SetConsumerServices(value []ConsumerService) Topic
+}
+
+// ConsumerService is a service that consumes the messages in a topic.
+type ConsumerService interface {
+	// ServiceID returns the service id of the consumer service.
+	ServiceID() services.ServiceID
+
+	// SetServiceID sets the service id of the consumer service.
+	SetServiceID(value services.ServiceID) ConsumerService
+
+	// ConsumptionType returns the consumption type of the consumer service.
+	ConsumptionType() ConsumptionType
+
+	// SetConsumptionType sets the consumption type of the consumer service.
+	SetConsumptionType(value ConsumptionType) ConsumerService
+
+	// String returns the string representation of the consumer service.
+	String() string
+}
+
+// Watch watches the updates of a topic.
+type Watch interface {
+	C() <-chan struct{}
+	Get() Topic
```

Comment:
As mentioned in a previous comment - if why shouldn't we allow for Get() to propagate errors? Especially if this interface is to be used in applications where there can be errors in retrieving values.

Reply:
This is to keep the convention as other Watches, like m3x or m3cluster, we could return `Topic, error` here, but user probably can't do much with the error update as well. I don't feel strongly either way.